model/legacy-private/src/main/java/org/keycloak/migration/migrators/MigrateTo22_0_0.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2023 Red Hat, Inc. and/or its affiliates
+ *  and other contributors as indicated by the @author tags.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package org.keycloak.migration.migrators;
+
+import org.jboss.logging.Logger;
+import org.keycloak.migration.ModelVersion;
+import org.keycloak.models.AuthenticationFlowModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.ModelException;
+import org.keycloak.models.RealmModel;
+import org.keycloak.representations.idm.RealmRepresentation;
+
+/**
+ * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
+ */
+public class MigrateTo22_0_0 implements Migration {
+
+    public static final ModelVersion VERSION = new ModelVersion("22.0.0");
+
+    public static final String HTTP_CHALLENGE_FLOW = "http challenge";
+
+    private static final Logger LOG = Logger.getLogger(MigrateTo22_0_0.class);
+
+    @Override
+    public void migrate(KeycloakSession session) {
+        session.realms().getRealmsStream().forEach(this::removeHttpChallengeFlow);
+    }
+
+    @Override
+    public void migrateImport(KeycloakSession session, RealmModel realm, RealmRepresentation rep, boolean skipUserDependent) {
+        removeHttpChallengeFlow(realm);
+    }
+
+    private void removeHttpChallengeFlow(RealmModel realm) {
+        AuthenticationFlowModel httpChallenge = realm.getFlowByAlias(HTTP_CHALLENGE_FLOW);
+        if (httpChallenge == null) return;
+
+        try {
+            realm.removeAuthenticationFlow(httpChallenge);
+
+            // Builtin "Http challenge" flow had subflow of this name, which should be removed as well
+            AuthenticationFlowModel subflow = realm.getFlowByAlias("Authentication Options");
```

Comment:
Doesn't it make more sense (and safer) to check if the authenticator is used within the flows rather than relying on the alias?

Reply:
The `Authentication Options` was subflow of the `Http Challenge` builtin flow. I cannot check for it's use with `KeycloakModelUtils.isFlowUsed` as it is just subflow, but I am checking that it's "builtin", which I supposed to be sufficient indication that it's the subflow of the `Http Challenge`.

I was considering another option - basically re-use same algorithm as used in `AuthenticationManagementResource.deleteFlow` to recursively delete the `Http Challenge` flow including it's subflow. Was about to move that stuff to `KEycloakModelUtils` to be able to use it from both services and migration code.

The small disadvantage of it are likely more DB queries as it needs to query executions etc, which looks like unnecessary DB queries during migration - especially in deployments with many realms. As we basically know that builtin flow `Http Challenge` has only 1 subflow called `Authentication Options`.

On the other hand, it's more proper approach to use `deleteFlow`, so I can change it if you prefer.

Reply:
TBH, I'm wondering if we should have migration at all but just include release notes.

We are adding code to something that is probably not used at all in the wild. And if so, it is supposed to break and force manual intervention based on the steps from the release notes.

Wdyt? Should be far simpler and less code to maintain and support.

Reply:
Well, I think that we definitely should have migration as otherwise people migrating from previous version will have `Http Challenge` flow in their realm, which would be just broken (as corresponding authenticators doesn't exists) and after click on this flow in the admin console, the error page would be shown. This looks like quite a buggy behaviour. The unused mess should be just removed IMO :-)